idx/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,441 @@
+	MaxBufferDocs  int
+	BufferDelayMax string
+)
+
+func ConfigSetup() {
+	esIdx := flag.NewFlagSet("elasticsearch-idx", flag.ExitOnError)
+	esIdx.BoolVar(&Enabled, "enabled", false, "")
+	esIdx.StringVar(&Index, "index", "metric", "Elasticsearch index name for storing metric index.")
+	esIdx.StringVar(&Hosts, "hosts", "localhost:9200", "comma separated list of elasticsearch address in host:port form")
+	esIdx.StringVar(&User, "user", "", "HTTP basic Auth username")
+	esIdx.StringVar(&Pass, "pass", "", "HTTP basic Auth password")
+	esIdx.StringVar(&RetryInterval, "retry-interval", "1h", "Interval to retry indexing Definitions in ES after failures")
+	esIdx.IntVar(&MaxConns, "max-conns", 20, "Max number of http conns in flight to ES servers at one time")
+	esIdx.IntVar(&MaxBufferDocs, "max-buffer-docs", 1000, "Max number of Docs to hold in buffer before forcing flush to ES")
+	esIdx.StringVar(&BufferDelayMax, "buffer-delay-max", "10s", "Max time to wait before flusing forcing flush to ES")
+	globalconf.Register("elasticsearch-idx", esIdx)
+}
+
+type RetryBuffer struct {
+	Index       *EsIdx
+	Defs        []schema.MetricDefinition
+	LastAttempt time.Time
+	done        chan struct{}
+	sync.Mutex
+}
+
+func NewRetryBuffer(index *EsIdx, interval time.Duration) *RetryBuffer {
+	r := &RetryBuffer{
+		Index:       index,
+		Defs:        make([]schema.MetricDefinition, 0),
+		LastAttempt: time.Now(),
+		done:        make(chan struct{}),
+	}
+	go r.run(interval)
+	return r
+}
+
+func (r *RetryBuffer) Stop() {
+	close(r.done)
+}
+
+func (r *RetryBuffer) Items() []schema.MetricDefinition {
+	r.Lock()
+	defs := r.items()
+	r.Unlock()
+	return defs
+}
+
+func (r *RetryBuffer) items() []schema.MetricDefinition {
+	defs := make([]schema.MetricDefinition, len(r.Defs))
+	copy(defs, r.Defs)
+	return defs
+}
+
+func (r *RetryBuffer) Retry(id string) {
+	def, err := r.Index.Get(id)
+	if err != nil {
+		log.Error(3, "Failed to get %s from Memory Index. %s", id, err)
+		return
+	}
+	r.Lock()
+	r.Defs = append(r.Defs, def)
+	r.Unlock()
+}
+
+func (r *RetryBuffer) retry() {
+	r.Lock()
+	defs := r.items()
+	r.Defs = r.Defs[:0]
+	r.Unlock()
+	if len(defs) == 0 {
+		log.Debug("retry buffer is empty")
+		return
+	}
+	for _, d := range defs {
+		if err := r.Index.BulkIndexer.Index(Index, "metric_index", d.Id, "", "", nil, d); err != nil {
+			log.Error(3, "Failed to add metricDef to BulkIndexer queue. %s", err)
+			r.Defs = append(r.Defs, d)
+			return
+		}
+	}
+}
+
+func (r *RetryBuffer) run(interval time.Duration) {
+	log.Debug("Starting RetryBuffer")
+	ticker := time.NewTicker(interval)
+	for {
+		select {
+		case <-r.done:
+			r.Lock()
+			log.Info("RetryBuffer stopping. %d items in cache have not been indexed.", len(r.Defs))
+			ticker.Stop()
+			r.Unlock()
+			return
+		case <-ticker.C:
+			r.retry()
+		}
+	}
+}
+
+// Implements the the "MetricIndex" interface
+type EsIdx struct {
+	memory.MemoryIdx
+	Conn        *elastigo.Conn
+	BulkIndexer *elastigo.BulkIndexer
```

Comment:
Why not?  I can think of lots of use cases where someone might want to extend the EsIdx and would want to make calls to ES.